### PR TITLE
definitions support specify the revision name

### PR DESCRIPTION
**Pull request by yangsoon**
**What this PR does / why we need it**:

Everytime you update the spec of deifnitions, KubeVela will automatic generated definitionRevision with name in
in the format of `definitionName@revisionNum`.  

Now, this PR supports you specify the revision name in annotation `definitionrevision.oam.dev/name`

```yaml
#worker.yaml
apiVersion: core.oam.dev/v1beta1
kind: ComponentDefinition
metadata:
  name: worker
  namespace: vela-system
  annotations:
    definition.oam.dev/description: "Describes long-running, scalable, containerized services that running at backend. They do NOT have network endpoint to receive external network traffic."
    # you can specify the revision name in annotations
    definitionrevision.oam.dev/name: "1.1.3"
spec:
  workload:
    definition:
      apiVersion: apps/v1
      kind: Deployment
  schematic:
    cue:
      template: |
        output: {
        	apiVersion: "apps/v1"
        	kind:       "Deployment"
        	spec: {
        		selector: matchLabels: {
        			"app.oam.dev/component": context.name
        		}

        		template: {
        			metadata: labels: {
        				"app.oam.dev/component": context.name
        			}
         ..........
```

```shell
kubectl apply -f worker.yaml
```

after apply the `worker` definition, KubeVela will create a definitionRevision named `worker-v1.1.3`

```shell
$ kubectl get definitionrevisions.core.oam.dev -A
NAMESPACE            NAME                                 REVISION   HASH               TYPE
vela-system          worker-v1.1.3                        1          a6d925a57ba3c297   Component
```

then you can specify the target componentDeifnition in your application

```yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: first-vela-appvprod
  namespace: default
spec:
  components:
    - name: test
      properties:
        image: crccheck/hello-world
      type: worker@v1.1.3
```

- [x] support named defRev
- [x] add test

### CODE REVIEW

File: apis/core.oam.dev/v1beta1/componentdefinition_types.go [hunk 1]
```diff
@@ -66,6 +66,8 @@ type ComponentDefinitionStatus struct {
 	// LatestRevision of the component definition
 	// +optional
 	LatestRevision *common.Revision `json:"latestRevision,omitempty"`
+	// MaxRevision record the max revision number
+	MaxRevision int64 `json:"max_revision"`
```
Comment (wonderflow) [inline apis/core.oam.dev/v1beta1/componentdefinition_types.go:70]: why do we need this field?

Comment (wonderflow) [reply apis/core.oam.dev/v1beta1/componentdefinition_types.go:70]: And this is a required field?

File: legacy/charts/vela-core-legacy/crds/core.oam.dev_componentdefinitions.yaml [hunk 2]
```diff
@@ -27,268 +27,532 @@ spec:
...
+            type: object
+          spec:
+            description: ComponentDefinitionSpec defines the desired state of ComponentDefinition
+            properties:
+              childResourceKinds:
+                description: ChildResourceKinds are the list of GVK of the child resources
+                  this workload generates
+                items:
+                  description: A ChildResourceKind defines a child Kubernetes resource
+                    kind with a selector
```
Comment (wonderflow) [inline legacy/charts/vela-core-legacy/crds/core.oam.dev_componentdefinitions.yaml:58]: why the crd has so much difference in this PR?

File: pkg/oam/util/helper_test.go [hunk 3]
```diff
@@ -2223,42 +2223,33 @@ func TestExtractRevisionNum(t *testing.T) {
...
 	}, {
 		defName:     "webservice@v10@v3",
-		wantRevName: "webservice@v10-v3",
-		hasError:    false,
+		wantRevName: "webservice-v10@v3",
 	}, {
 		defName:     "@v10",
-		wantRevName: "",
-		hasError:    true,
+		wantRevName: "@v10",
```
Comment (wonderflow) [inline pkg/oam/util/helper_test.go:2247]: this is still error case. Please follow K8s name convention for validation.

File: pkg/oam/util/helper.go [hunk 4]
```diff
@@ -367,32 +367,31 @@ func GetCapabilityDefinition(ctx context.Context, cli client.Reader, definition
...
+	splits := strings.Split(definitionName, "@v")
+	if len(splits) == 1 || len(splits[0]) == 0 {
+		return definitionName
 	}
-	return fmt.Sprintf("%s-v%d", defName, revNum), nil
+
+	defName := splits[0]
+	revisionName := strings.TrimPrefix(definitionName, fmt.Sprintf("%s@v", defName))
+
+	return fmt.Sprintf("%s-v%s", defName, revisionName)
```
Comment (wonderflow) [inline pkg/oam/util/helper.go:394]: you should validate the final name

Comment (wonderflow) [reply pkg/oam/util/helper.go:394]: or add validation in webhook, then there's no need

Comment (yangsoon) [reply pkg/oam/util/helper.go:394]: sure, we must check the final name of definitionRevision, i will validate the name in webhook

File: legacy/charts/vela-core-legacy/crds/core.oam.dev_componentdefinitions.yaml (hunk 2 above)
Comment (yangsoon) [reply legacy/charts/vela-core-legacy/crds/core.oam.dev_componentdefinitions.yaml]: fix it

**Review by wonderflow — APPROVED** [review]

**Pull request merged**